### PR TITLE
Fix position resize calculations: average execution price and closed DLC channel settlement amount

### PR DESCRIPTION
**Pull request by holzeis**
We must not change the average entry price if the position is reduced as the user is not buying/selling more of the contract, but selling/buying it.

This resulted into the incorrect pnl calculation in case the position was reduced.

### CODE REVIEW

File: coordinator/src/node/resize.rs [hunk 1]
```diff
@@ -340,6 +340,12 @@ fn compute_average_execution_price(
 ) -> Decimal {
     let starting_average_execution_price = decimal_from_f32(starting_average_execution_price);
 
+    if starting_direction != trade_direction {
```
Comment (bonomat) [inline coordinator/src/node/resize.rs:343]: If he is selling more than he had before the execution price should update, i.e. if he had 20 contracts long and sold 30 short he is now 10 negative at the new price.

**Review by bonomat — COMMENTED** [review]
Can you add tests for this please?

File: coordinator/src/node/resize.rs (hunk 1 above)
Comment (holzeis) [reply coordinator/src/node/resize.rs:343]: https://github.com/get10101/10101/pull/1628#issue-2006431156

Comment (holzeis) [reply coordinator/src/node/resize.rs:343]: > If he is selling more than he had before the execution price should update, i.e. if he had 20 contracts long and sold 30 short he is now 10 negative at the new price.

How is that average_execution_price calculated? In this context we only have on order matched by the bid / ask price, but we would actually need two orders don't we? one for the ask and one for the bid (and vice versa)

File: mobile/native/src/trade/position/handler.rs [hunk 2]
```diff
@@ -39,34 +39,47 @@ pub async fn trade(filled: FilledWith) -> Result<()> {
...
-    {
-        let position_contracts_relative = compute_relative_contracts(*quantity, *direction);
-        let trade_contracts_relative =
-            compute_relative_contracts(trade_params.quantity, trade_params.direction);
-
-        if *contract_symbol == trade_params.contract_symbol
-            && position_contracts_relative + trade_contracts_relative != Decimal::ZERO
-        {
-            set_position_state(PositionState::Resizing)?;
+    let execution_price = match get_positions()?.first() {
```
Comment (luckysori) [inline mobile/native/src/trade/position/handler.rs:50]: 🤔 I don't get why we need to compute an execution price. Surely, the execution price should be the one from the trade params, no matter the state of the current position. Examples:

1. If we are 5 long and we go 5 long at price X, we are now 10 long with an average execution price that takes into account price X.
2. If we are 5 long and we go 2 short, we are now 3 long and the PNL uses the trade execution price as a closing price. Importantly the average execution price for the position should not change in such a scenario. I need to check that.
3. If we are 5 long and we got 5 short, we no longer have a position and the PNL uses the trade execution price as the closing price.
4. If we are 5 long and we go 10 short, we are now 5 short with the trade execution price as the new average entry price. The PNL for the closed long position is calculated using the trade execution price as the closing price.

I don't think we want to mess with the trade execution price we record in the database in any of these scenarios.

Comment (luckysori) [reply mobile/native/src/trade/position/handler.rs:50]: > I need to check that.

I found one UI bug already. We were wrongly updating the average execution price of the position when partially reducing the position.

**Comment by luckysori** [comment]
@holzeis: I've updated the PR with my take on this. Testing the scenario where the position is first reduced and then closed, the PNL changes correctly.

**Comment by holzeis** [comment]
> @holzeis: I've updated the PR with my take on this. Testing the scenario where the position is first reduced and then closed, the PNL changes correctly.

Awesome, looks great 🚀

**Review by luckysori — APPROVED** [review]
What a great PR.

**Pull request merged**